Always log debug information for fee calculation in CreateTransaction

morcos opened the pull request:
This is a first pass at always logging a single line of information whenever CreateTransaction is invoked which will help debug any problems in fee calculation or estimation.

Built on top of #10199 and #10283, only the last commit is new.

morcos committed:
Output Fee Estimation Calculations in CreateTransaction

jonasschnelli commented on src/wallet/feebumper.cpp:
```diff
@@ -160,7 +160,7 @@ CFeeBumper::CFeeBumper(const CWallet *pWallet, const uint256 txidIn, int newConf
         // if user specified a confirm target then don't consider any global payTxFee
         if (specifiedConfirmTarget) {
-            nNewFee = CWallet::GetMinimumFee(maxNewTxSize, newConfirmTarget, mempool, ::feeEstimator, true);
+            nNewFee = CWallet::GetMinimumFee(maxNewTxSize, newConfirmTarget, mempool, ::feeEstimator, nullptr, true);
```
Would logging the fee estimation details during bumpfee (all call to `GetMinimumFee`) be possible?

morcos replied:
I think this is possible, but it wasn't clear to me we wanted it here.
Perhaps that could be added later

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -2777,23 +2787,32 @@ CAmount CWallet::GetRequiredFee(unsigned int nTxBytes)
-        if (nFeeNeeded == 0)
+        if (nFeeNeeded == 0) {
             nFeeNeeded = fallbackFee.GetFee(nTxBytes);
+            if (feeCalc) feeCalc->reason = FeeReason::FALLBACK;
```
supernit: use brackets `if (feeCalc) { feeCalc->reason = FeeReason::FALLBACK; }`

sipa replied:
No. From https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md:
```
If an if only has a single-statement then-clause, it can appear on the same line as the if, without braces. In every other case, braces are required, and the then and else clauses must appear correctly indented on a new line.
```

jonasschnelli replied:
Okay. Fair point.

morcos commented:
rebased

laanwj commented on src/policy/fees.h:
```diff
@@ -74,6 +74,33 @@ enum FeeEstimateHorizon {
 };
 
+/* Enumeration of reason for returned fee estimate */
+enum FeeReason {
```
shouldn't this be "enum class" to avoid the enum values from leaking into global scope?

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -2734,6 +2735,15 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
         }
     }
+
+    LogPrintf("Fee Calculation: Fee:%s Bytes:%u Tgt:%d (requested %d) Reason:\"%s\" Decay %.5f: Estimation: (%g - %g) %.2f%% %.1f/(%.1f %d mem %.1f out) Fail: (%g - %g) %.2f%% %.1f/(%.1f %d mem %.1f out)\n",
```
Nice! That should give a lot of info for troubleshooting this.

jnewbery replied:
Should this be `LogPrint(BCLog::ESTIMATEFEE...` (ie only log if estimatefee logs are enabled)?

morcos replied:
It's meant to get logged even if you have not selected any debug log categories

ryanofsky reviewed (commented):
utACK d0fb0f69b91539166de98820251d3ac047be0de0, though needs rebase (conflicted with #10390).

morcos commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -679,7 +679,7 @@ void SendCoinsDialog::updateSmartFeeLabel()
                                                                 std::max(feeRate.GetFeePerK(), CWallet::GetRequiredFee(1000))) + "/kB");
         ui->labelSmartFee2->hide();
-        ui->labelFeeEstimation->setText(tr("Estimated to begin confirmation within %n block(s).", "", estimateFoundAtBlocks));
+        ui->labelFeeEstimation->setText(tr("Estimated to begin confirmation within %n block(s).", "", feeCalc.desiredTarget));
```
damn, i made that same bug elsewhere.  thx

morcos commented:
Addressed comments and squashed/rebased 
[debugEstimates.ver2](https://github.com/morcos/bitcoin/commits/debugEstimates.ver2) --> [debugEstimates.ver2.squash](https://github.com/morcos/bitcoin/commits/debugEstimates.ver2.squash)  --> [debugEstimates.ver2.rebased](https://github.com/morcos/bitcoin/commits/debugEstimates.ver2.rebased)

ryanofsky commented on src/policy/fees.h:
```diff
@@ -74,6 +74,22 @@ enum FeeEstimateHorizon {
+    MAXTXFEE,
+};
+
+const std::string StringForFeeReason(FeeReason reason);
```
I don't think there's any benefit to making this const (cases where you do want this are pretty rare https://stackoverflow.com/questions/8716330/purpose-of-returning-by-const-value).

morcos replied:
done

ryanofsky replied:
Const is actually still here (you removed it from the definition but not declaration)

ryanofsky reviewed (commented):
utACK 436207ef833dfdd6dc2e664799a68516cb3249a8. Looks good. Changes since previous review were rebase, cleaned up feereason enum/strings, fixed ui target string.

morcos commented:
addressed nit

ryanofsky reviewed (commented):
utACK 7c287a7b5df6f33f3b7723496590e72cdd7fed2e. Only change is removed const.

morcos commented:
oops, actually addressed nit this time.

ryanofsky reviewed (commented):
utACK 1bebfc8d3aa615dfdd0221f21b87319e36821b71, const is gone

laanwj commented:
utACK 1bebfc8
